### PR TITLE
Accept stage codes

### LINKED ISSUE #110: (URGENT) Stages (and abbreviations, and codes) must be validated and throw error when wrong
Stage 40 => DIS:
```ruby
irb(main):004:0> Pubid::Iso::Identifier.new(publisher: "ISO", number: 1000, stage: 40).to_s
```
Should be "ISO/DIS 1000".

```ruby
irb(main):005:0> Pubid::Iso::Identifier.new(publisher: "ISO", number: 1000, stage: "40").to_s
```
Should be "ISO/DIS 1000".

```ruby
irb(main):006:0> Pubid::Iso::Identifier.new(publisher: "ISO", number: 1000, stage: "40.40").to_s
```
Should be "ISO/DIS 1000".

~~This should throw an error, because `WDD` is not a defined stage.~~

```ruby
irb(main):008:0> Pubid::Iso::Identifier.new(publisher: "ISO", number: 1000, stage: :WDD).to_s
pubid-iso-0.1.12/lib/pubid/iso/stage.rb:27:in `initialize': WDD is not valid stage (Pubid::Iso::Errors::CodeInvalidError)
```

Maybe `StageInvalidError`?

### LINKED ISSUE #114: Implement accurate numeric stage codes to abbreviations conversion
Now "20.60" returns "AWI" instead of "WD"
"10.99" returns "NP" instead of "AWI"

https://github.com/metanorma/pubid-iso/issues/48#issuecomment-1209330306

| International harmonized stage code | Development stage
| -- | -- |
| 00.00, 00.20, 00.60, 00.98, 00.99 | `:pwi` |
| 10.00, 10.20, 10.60, 10.92, 10.98 | `:np` |
| 10.99 | `:awi` |
| 20.00 | `:awi` |
| 20.20, 20.60, 20.98, 20.99 | `:wd` |
| 30.00, 30.20, 30.60, 30.92, 30.98, 30.99 | `:cd` |
| 40.00, 40.20, 40.60, 40.92, 40.93, 40.98, 40.99 | `:dis` |
| 50.00, 50.20, 50.60, 50.92, 50.98, 50.99 | `:fdis` (default) or `:prf` |
| 60.00 | `:prf` |
| 60.60 | `:published` |
| 90.20, 90.60, 90.92, 90.93, 90.99 | `:published` |
| 95.20, 95.60, 95.92, 95.99  | `:published` |

NOTE: In ISO, the only conflicting stage code is 50.00 to 50.99. 

1. We need to implement a class called `HarmonizedStageCode`, it is composed of `stage_code: NN, substage_code: MM`. 
2. In `Pubid::Iso`, we need to have a class `DevelopmentStage` that contains a `HarmonizedStageCode` and the development stage (e.g. `:pwi`, `:np` ...)
3. The rendering of the `DevelopmentStage` in the output differs per document type.

_Originally posted by @ronaldtse in https://github.com/metanorma/pubid-iso/issues/48#issuecomment-1234970429_

### PR DESCRIPTION
**Pull request by mico**
closes #110 
closes #114

### CODE REVIEW

**Review by ronaldtse — CHANGES_REQUESTED** [review]

File: lib/pubid/iso/stage.rb [hunk 1]
```diff
@@ -13,21 +13,34 @@ class Stage
...
-        raise Errors::CodeInvalidError, "#{abbr} is not valid stage" unless STAGES.key?(abbr.to_sym)
+        raise Errors::StageInvalidError, "#{abbr} is not valid stage" unless STAGES.key?(abbr.to_sym)
 
         @harmonized_code ||= HarmonizedStageCode.new(*STAGES[abbr.to_sym].split("."))
       end
     end
+
+    def self.parse(stage)
+      if /\A[\d.]+\z/.match?(stage)
+        Stage.new(harmonized_code: stage)
```
[the remark this answers is not in the document]
Comment (ronaldtse) [reply lib/pubid/iso/stage.rb:61]: @mico we are almost there. There are some code combinations that are not allowed, that's why I want you to encode ALL the valid codes. For example, 10.92, 10.93, 20.92 are invalid. Also 60.20, 60.9X, 95.00 are invalid.

[the remark this answers is not in the document]
Comment (ronaldtse) [reply lib/pubid/iso/stage.rb:61]: Hmm it looks like you do have tests for the invalid codes. How are you rejecting those?

[the remark this answers is not in the document]
Comment (mico) [reply lib/pubid/iso/stage.rb:61]: > Hmm it looks like you do have tests for the invalid codes. How are you rejecting those?

I'm using DESCRIPTIONS list from https://github.com/metanorma/pubid-iso/blob/main/lib/pubid/iso/harmonized_stage_code.rb#L6

[the remark this answers is not in the document]
Comment (ronaldtse) [reply lib/pubid/iso/stage.rb:61]: Thank you! That's perfect!

**Pull request merged**